File: moling-server/app/router/project.py
"""墨灵 (Moling) — Project API Router.

Provides endpoints for project CRUD operations.
"""

from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db, get_current_user
from app.schemas.project import CreateProjectReq, ProjectResp, ProjectStatsResp, UpdateProjectReq
from app.service import project_service

router = APIRouter(tags=["projects"])


@router.post("", response_model=ProjectResp, status_code=201)
async def create_project(
    req: CreateProjectReq,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> ProjectResp:
    """Create a new project."""
    return await project_service.create_project(db, current_user.id, req)


@router.get("", response_model=dict)
async def list_projects(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Page size"),
    status: Optional[str] = Query(None, description="Filter by status"),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> dict:
    """List projects with pagination."""
    return await project_service.list_projects(db, current_user.id, page, page_size, status)


@router.get("/stats", response_model=ProjectStatsResp)
async def get_project_stats(
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> ProjectStatsResp:
    """Get project statistics for current user."""
    return await project_service.get_project_stats(db, current_user.id)


@router.get("/{project_id}/stats", response_model=dict)
async def get_single_project_stats(
    project_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> dict:
    """获取单个项目的统计信息。"""
    project = await project_service.get_project(db, current_user.id, project_id)
    return {
        "project_id": project.id,
        "title": project.title,
        "total_chapters": project.chapters,
        "total_words": project.word_count,
        "status": project.status,
    }


@router.get("/{project_id}", response_model=ProjectResp)
async def get_project(
    project_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> ProjectResp:
    """Get single project by ID."""
    return await project_service.get_project(db, current_user.id, project_id)


@router.put("/{project_id}", response_model=ProjectResp)
async def update_project(
    project_id: int,
    req: UpdateProjectReq,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> ProjectResp:
    """Update project by ID."""
    return await project_service.update_project(db, current_user.id, project_id, req)


@router.delete("/{project_id}", status_code=204)
async def delete_project(
    project_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> None:
    """Delete project by ID."""
    await project_service.delete_project(db, current_user.id, project_id)


@router.get("/{project_id}/suggestions", response_model=dict)
async def get_project_suggestions(
    project_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> dict:
    """获取项目的创作建议（基于四库分析）。"""
    suggestions = await project_service.get_suggestions(
        db, current_user.id, project_id
    )
    return suggestions



# ============ Project Cards History ============

@router.get("/{project_id}/cards/history", response_model=list[dict])
async def get_project_card_history(
    project_id: int,
    chapter_id: Optional[int] = Query(None, description="按章节过滤"),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> list[dict]:
    """获取项目的抽卡历史。"""
    from app.service.card_service import card_service
    return await card_service.get_draw_history(db, current_user.id, project_id, chapter_id)

File: moling-server-rs/crates/moling-api/src/routes/project.rs
//! Project routes — CRUD, stats, and suggestions.

use axum::{
    extract::{Path, Query, State},
    routing::{get, post},
    Json, Router,
};
use moling_core::error::{AppError, AppResult};
use moling_core::types::Pagination;
use moling_db::dao::project_dao::ProjectDao;
use moling_auth::CurrentUser;
use sea_orm::Set;
use crate::state::AppState;
use crate::types::*;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", post(create_project).get(list_projects))
        .route("/stats", get(get_stats))
        .route("/{project_id}", get(get_project).put(update_project).delete(delete_project))
        .route("/{project_id}/stats", get(get_single_stats))
        .route("/{project_id}/suggestions", get(get_suggestions))
}

/// POST /projects — create a new project.
#[utoipa::path(
    post,
    path = "/api/v1/projects",
    request_body = CreateProjectReq,
    responses(
        (status = 200, description = "Project created successfully", body = ProjectResp),
        (status = 401, description = "Not authenticated")
    )
)]
pub async fn create_project(
    State(state): State<AppState>,
    user: CurrentUser,
    Json(req): Json<CreateProjectReq>,
) -> AppResult<Json<ProjectResp>> {
    let dao = ProjectDao;
    let model = moling_db::entities::project::ActiveModel {
        user_id: Set(user.user_id.to_string()),
        title: Set(req.title),
        author: Set(req.author.unwrap_or_default()),
        genre: Set(req.genre.unwrap_or_default()),
        synopsis: Set(req.synopsis),
        worldview: Set(req.worldview),
        protagonist: Set(req.protagonist),
        style: Set(req.style),
        target_words: Set(req.target_words),
        frequency: Set(req.frequency),
        ..Default::default()
    };
    let p = dao.create(&state.db, model).await?;
    Ok(Json(to_resp(p)))
}

/// GET /projects — list user's projects with pagination.
#[utoipa::path(
    get,
    path = "/api/v1/projects",
    params(
        ("page" = Option<u32>, Query, description = "Page number (default: 1)"),
        ("page_size" = Option<u32>, Query, description = "Page size (default: 20)"),
    ),
    responses(
        (status = 200, description = "Paginated list of projects"),
        (status = 401, description = "Not authenticated")
    )
)]
pub async fn list_projects(
    State(state): State<AppState>,
    user: CurrentUser,
    Query(pg): Query<PaginationQuery>,
) -> AppResult<Json<PaginatedList<ProjectResp>>> {
    let dao = ProjectDao;
    let pagination = Pagination { page: pg.page, page_size: pg.page_size };
    let (items, total) = dao.find_by_user(&state.db, &user.user_id.to_string(), &pagination).await?;
    Ok(Json(PaginatedList {
        items: items.into_iter().map(to_resp).collect(),
        total,
        page: pg.page,
        page_size: pg.page_size,
    }))
}

/// GET /projects/{project_id} — get project details.
#[utoipa::path(
    get,
    path = "/api/v1/projects/{project_id}",
    params(
        ("project_id" = i32, Path, description = "Project ID"),
    ),
    responses(
        (status = 200, description = "Project details", body = ProjectResp),
        (status = 401, description = "Not authenticated"),
        (status = 403, description = "Access denied"),
        (status = 404, description = "Project not found")
    )
)]
pub async fn get_project(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(project_id): Path<i32>,
) -> AppResult<Json<ProjectResp>> {
    let dao = ProjectDao;
    let p = dao.find_by_id(&state.db, project_id).await?.ok_or_else(AppError::project_not_found)?;
    if p.user_id != user.user_id.to_string() {
        return Err(AppError::project_access_denied());
    }
    Ok(Json(to_resp(p)))
}

/// PUT /projects/{project_id} — update project.
#[utoipa::path(
    put,
    path = "/api/v1/projects/{project_id}",
    params(
        ("project_id" = i32, Path, description = "Project ID"),
    ),
    request_body = UpdateProjectReq,
    responses(
        (status = 200, description = "Project updated", body = ProjectResp),
        (status = 401, description = "Not authenticated"),
        (status = 403, description = "Access denied"),
        (status = 404, description = "Project not found")
    )
)]
pub async fn update_project(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(project_id): Path<i32>,
    Json(req): Json<UpdateProjectReq>,
) -> AppResult<Json<ProjectResp>> {
    use sea_orm::{ActiveModelTrait, IntoActiveModel};
    let dao = ProjectDao;
    let p = dao.find_by_id(&state.db, project_id).await?.ok_or_else(AppError::project_not_found)?;
    if p.user_id != user.user_id.to_string() {
        return Err(AppError::project_access_denied());
    }
    let mut active = p.into_active_model();
    if let Some(v) = req.title { active.title = Set(v); }
    if let Some(v) = req.author { active.author = Set(v); }
    if let Some(v) = req.genre { active.genre = Set(v); }
    if let Some(v) = req.synopsis { active.synopsis = Set(Some(v)); }
    if let Some(v) = req.worldview { active.worldview = Set(Some(v)); }
    if let Some(v) = req.protagonist { active.protagonist = Set(Some(v)); }
    if let Some(v) = req.style { active.style = Set(Some(v)); }
    if let Some(v) = req.target_words { active.target_words = Set(Some(v)); }
    if let Some(v) = req.frequency { active.frequency = Set(Some(v)); }
    let updated = active.update(&state.db).await.map_err(|_| AppError::internal("Update failed".to_owned()))?;
    Ok(Json(to_resp(updated)))
}

/// DELETE /projects/{project_id} — delete a project.
#[utoipa::path(
    delete,
    path = "/api/v1/projects/{project_id}",
    params(
        ("project_id" = i32, Path, description = "Project ID"),
    ),
    responses(
        (status = 204, description = "Project deleted"),
        (status = 401, description = "Not authenticated"),
        (status = 403, description = "Access denied"),
        (status = 404, description = "Project not found")
    )
)]
pub async fn delete_project(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(project_id): Path<i32>,
) -> AppResult<axum::http::StatusCode> {
    let dao = ProjectDao;
    let p = dao.find_by_id(&state.db, project_id).await?.ok_or_else(AppError::project_not_found)?;
    if p.user_id != user.user_id.to_string() {
        return Err(AppError::project_access_denied());
    }
    dao.soft_delete(&state.db, project_id).await?;
    Ok(axum::http::StatusCode::NO_CONTENT)
}

async fn get_stats(
    State(state): State<AppState>,
    user: CurrentUser,
) -> AppResult<Json<ProjectStatsResp>> {
    let dao = ProjectDao;
    let count = dao.count_by_user(&state.db, &user.user_id.to_string()).await?;
    Ok(Json(ProjectStatsResp { total_projects: count, total_words: 0, total_chapters: 0 }))
}

async fn get_single_stats(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(project_id): Path<i32>,
) -> AppResult<Json<SingleProjectStatsResp>> {
    let dao = ProjectDao;
    let p = dao.find_by_id(&state.db, project_id).await?.ok_or_else(AppError::project_not_found)?;
    if p.user_id != user.user_id.to_string() {
        return Err(AppError::project_access_denied());
    }
    Ok(Json(SingleProjectStatsResp {
        project_id: p.id,
        title: p.title,
        total_chapters: 0,
        total_words: p.word_count,
        status: p.status,
    }))
}

async fn get_suggestions(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(project_id): Path<i32>,
) -> AppResult<Json<ProjectSuggestionResp>> {
    let dao = ProjectDao;
    let p = dao.find_by_id(&state.db, project_id).await?.ok_or_else(AppError::project_not_found)?;
    if p.user_id != user.user_id.to_string() {
        return Err(AppError::project_access_denied());
    }
    let _ = p; // suggestions logic will be added in service phase
    Ok(Json(ProjectSuggestionResp { suggestions: vec![] }))
}

fn to_resp(p: moling_db::entities::project::Model) -> ProjectResp {
    ProjectResp {
        id: p.id,
        user_id: p.user_id,
        title: p.title,
        author: p.author,
        genre: p.genre,
        synopsis: p.synopsis,
        worldview: p.worldview,
        protagonist: p.protagonist,
        style: p.style,
        word_count: p.word_count,
        target_words: p.target_words,
        status: p.status,
        creation_mode: p.creation_mode,
        created_at: p.created_at.into(),
        updated_at: p.updated_at.into(),
    }
}
